services/webnn/ort/graph_builder_ort.cc

```diff
@@ -258,6 +259,658 @@ GraphBuilderOrt::GraphBuilderOrt(
+      output_ids = {softsign->output_operand_id};
+      break;
+    }
+    case mojom::Operation::Tag::kSplit: {
+      const auto& split = operation->get_split();
+      input_ids = {split->input_operand_id};
+      output_ids = {split->output_operand_ids};
+      break;
+    }
+    case mojom::Operation::Tag::kTanh: {
+      const auto& tanh = operation->get_tanh();
+      input_ids = {tanh->input_operand_id};
+      output_ids = {tanh->output_operand_id};
+      break;
+    }
+    case mojom::Operation::Tag::kTile: {
+      const auto& tile = operation->get_tile();
+      input_ids = {tile->input_operand_id};
+      output_ids = {tile->output_operand_id};
+      break;
+    }
+    case mojom::Operation::Tag::kTranspose: {
+      const auto& transpose = operation->get_transpose();
+      input_ids = {transpose->input_operand_id};
+      output_ids = {transpose->output_operand_id};
+      break;
+    }
+    case mojom::Operation::Tag::kTriangular: {
+      const auto& triangular = operation->get_triangular();
+      input_ids = {triangular->input_operand_id};
+      output_ids = {triangular->output_operand_id};
+      break;
+    }
+    case mojom::Operation::Tag::kWhere: {
+      const auto& where = operation->get_where();
+      input_ids = {where->condition_operand_id, where->true_value_operand_id,
+                   where->false_value_operand_id};
+      output_ids = {where->output_operand_id};
+      break;
+    }
+  }
+}
+
+// The struct contains the information of graph fusion. In `CreateAndBuild`
+// method, when going through all operations to add each operation into the
+// final graph, this struct will be used for graph fusion.
+struct GraphFusionInfo {
+  // A map of all operations that can be fused into the following matmul
+  // using the fusiable operation's output operand id as the key.
+  std::map<uint64_t, raw_ptr<const mojom::Operation, CtnExperimental>>
+      output_id_to_fusible_matmulnbits_map;
+
+  // A set of all operations in `mojom::GraphInfo` which can be fused into
+  // another operation. No ORT operator node will be created for operations
+  // in this set.
+  std::unordered_set<const mojom::Operation*> fusible_operations_set;
+};
+
+// The method gets the graph fusion information from `mojom::GraphInfo`, based
+// on that the `operations` in `mojom::GraphInfo` have been in topological
+// order which means if operation 'j' depends on 'i', 'i' must appear before
+// 'j'.
+GraphFusionInfo GetGraphFusionInfo(const mojom::GraphInfo& graph_info) {
+  // If it's disabled, just return empty 'GraphFusionInfo' object which means no
+  // graph fusion will be applied since currently we only enable matmulnbits
+  // fusion.
+  if (!base::CommandLine::ForCurrentProcess()->HasSwitch(
+          switches::kWebNNOrtApplyMatMulNBitsFusion)) {
+    return GraphFusionInfo();
+  }
+
+  GraphFusionInfo graph_fusion_info;
+
+  // A map of all matmul operations in `mojom::GraphInfo` using matmul's input
```

Comment:
2nd input? Does the 2nd input really matter? Can the 1st input be supported as well?